### PR TITLE
Updating darc dependencies and enabling closure

### LINKED ISSUE #44078: Update Assembly Versions to 6.0.0.0
- [x] Update assemblyVersions to 6.0.0.0
- [x] Package index needs to be updated for assemblies which are added in the timeframe between the 6.0 snap and this change.
- [ ] Reenable the validation after a round trip change

cc @ericstj @ViktorHofer @safern

### PR DESCRIPTION
**Pull request by Anipik**
Fixes https://github.com/dotnet/runtime/issues/44078
Continuation of https://github.com/dotnet/runtime/pull/45386

The other pr got automatically closed on a force push, there were some trouble while resolving merge conflicts with the recent coreclr change

### CODE REVIEW

File: src/libraries/Directory.Build.targets
Comment (safern) [inline src/libraries/Directory.Build.targets:295]: I think we should instead figure out why we need all this. This is causing issues depending on the SDK's installed on the machine. https://github.com/dotnet/runtime/issues/45755

File: src/mono/Directory.Build.targets
Comment (ViktorHofer) [inline src/mono/Directory.Build.targets:19]: Assuming that these item groups are all identical across libraries, mono and installer, could we move them into a separate props file and import it conditionally in the places where we need it? The condition could we written in a way that the file is automatically imported when we use a tfm that the SDK doesn't yet support.

File: src/libraries/Directory.Build.targets
Comment (ViktorHofer) [reply src/libraries/Directory.Build.targets:295]: These items need to be set for any project that doesn't set `DisableImplicitFrameworkReferences` to true and uses a `TargetFramework` that the SDK doesn't understand yet.

Comment (ViktorHofer) [reply src/libraries/Directory.Build.targets:295]: I agree with Santi that we should not define them globally but only where we need them in a sharable manner.

File: eng/targetingpacks.targets
Comment (ViktorHofer) [inline eng/targetingpacks.targets:7]: @ericstj @Anipik why do we need to set this? It this something that changed recently in the SDK? Does the SDK error when using a local, live built targeting pack, the way we do? Assuming it doesn't as we would have hit this when we upgraded the SDK to 5.0.100.

Comment (ericstj) [reply eng/targetingpacks.targets:7]: The SDK doesn't download the targeting pack because of EnableTargetingPackDownload.  It does still check that the ref-pack is available.  In 5.0, the value of the ref-pack used was always lower than the one in the SDK, so the SDK would permit roll forward to its own copy (only to be later replaced by these targets), but avoiding the error.  Now we're trying to use a 5.0 GA SDK with a 6.0 ref pack, so it cannot rollforward and emits the error.  This suppresses that error.

Comment (ViktorHofer) [reply eng/targetingpacks.targets:7]: > In 5.0, the value of the ref-pack used was always lower than the one in the SDK, so the SDK would permit roll forward to its own copy

What do you mean by value of the ref-pack? Can you please elaborate a bit on this by providing examples of which versions were used in 5.0?

Comment (ericstj) [reply eng/targetingpacks.targets:7]: `KnownFrameworkReference` defines which version of the ref-pack is used for a given TFM (RuntimeFrameworkVersion can also override this).  Since the SDK doesn't support 6.0 yet, we're adding that support by adding a `KnownFrameworkReference` reference for it and pointing it at the 6.0 ref packs.  

In most of 5.0 we didn't override this, we just used the built in `KnownFrameworkReference` from the SDK which always had a version that match the ref-pack it redistributed.  Some projects did set RuntimeFrameworkVersion (which could be higher than the SDK) but those projects would just download the version they requested.  Early in 5.0 when we were using a pre-5.0 SDK we didn't have to worry about these targetingpack.targets that disable the download, because we weren't trying to replace the SDK's refpack resolution with our own, but I suspect we'll run into this same problem every time we try to use an SDK to target a new framework before we've moved to using an SDK that officially supports it.

Normally this "just works" for a project as the SDK will acquire that ref-pack via PackageDownload.  This isn't working for the library projects that use our own live-built refpack.  As I mentioned these targets disabled that acquisition, however they didn't disable the checking of presence of the refpack : https://github.com/dotnet/sdk/blob/44f381b62d466565639d51847c9127afbe7062a9/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets#L315-L346.  The SDK ends up checking the packs directory for this ref pack, doesn't find it and isn't permitted to download it, so it raises an error.  

Another version of this fix would be to disable that target completely, or make sure that target was probing our artifacts directory for the ref-pack rather than us just [overriding](https://github.com/dotnet/runtime/blob/96d7aecbf4b616002a6d0815cdbf125bc929012e/eng/targetingpacks.targets#L65-L80) its output.  It seems to me like given the current design of disabling the restore and overriding the result of ResolveTargetingPackAssets, we should also not care if ResolveTargetingPackAssets couldn't find it since we just rewrite its output to point to our built refpack.

Comment (ViktorHofer) [reply eng/targetingpacks.targets:7]: Thanks for the thorough explanation. I'm still confused why this regressed with the SDK update to 5.0.100 as we were already using the 6.0 TFM. I'm curious if this problem already existed before the SDK update but just didn't pop up as the targeting packs were already in the cache.

Comment (safern) [reply eng/targetingpacks.targets:7]: > I'm curious if this problem already existed before the SDK update but just didn't pop up as the targeting packs were already in the cache.

I believe this is happening when we updated the SDK to 5.0.100 because that one has the 5.0.0 targeting pack, however the `MicrosoftNetCoreAppVersion` which we use to specify the targeting pack version for the `net6.0` known framework reference is pointing to 5.0.0-rc1.* 

We should probably update as part of this PR the `MicrosoftNetCoreAppVersion` to `5.0.0` and also in any other places where we reference the `Microsoft.NETCore.App` version `5.0.0-rc.1.20451.14`.

Comment (ericstj) [reply eng/targetingpacks.targets:7]: Right, it was working because we were still using the 5.0 RC ref pack reference for net6.0, which the SDK was rolling forward to its own 5.0.0 ref pack.  😖

Comment (ViktorHofer) [reply eng/targetingpacks.targets:7]: > We should probably update as part of this PR the MicrosoftNetCoreAppVersion to 5.0.0 and also in any other places where we reference the Microsoft.NETCore.App version 5.0.0-rc.1.20451.14.

This PR is actually already updating that property, but to a 6.0.0-alpha version which confuses me. Why would we use a pre-built 6.0 reference anywhere instead of the live targeting pack for both the runtime and the apphost?

File: src/installer/Directory.Build.targets
Comment (ViktorHofer) [inline src/installer/Directory.Build.targets:166]: Doesn't this whole block suggest that we would use a a 5.0.0 targeting pack when compiling against net6.0 in installer? This doesn't make any sense to me but I hope that I'm misunderstanding something here.

Comment (ericstj) [reply src/installer/Directory.Build.targets:166]: I suspect this is due to the installer tests saying they "target 6.0" but aren't set up correctly to use the live bits nor run on the them either?  @Anipik mentioned something about this to me yesterday when I had the same question

File: eng/targetingpacks.targets
Comment (ericstj) [reply eng/targetingpacks.targets:7]: Certainly the 6.0 version is better than 5.0-rc for representing 6.0.  Imagine the mode project's are in once we actually get an LKG SDK with 6.0 support.  They'll be using some LKG 6.0 ref-pack.

Presumably there are projects in the repo that are needing to represent themselves as "net6.0" (for example in a package) but aren't set up to build against the live ref-pack that is built (we know of some in installer, not sure if there are any in coreclr too).  I agree we should weed out those projects and fix them: left over work from runtime consolidation.  Not sure if this PR is the right place for it, as this change is blocking the ingestion to SDK.

File: src/installer/Directory.Build.targets
Comment (ViktorHofer) [reply src/installer/Directory.Build.targets:166]: Shouldn't they then instead use `$(NetCoreAppToolCurrent)` which is always in sync with the sdk? Either that or making them dependent on the live built targeting pack which I assume we don't want.

File: src/libraries/Directory.Build.targets
Comment (ericstj) [reply src/libraries/Directory.Build.targets:295]: I think we need it in all of Libraries in order to target 6.0.  In libraries it should never cause the error (see my thread above where we set the property to disable that error).

Outside of libraries I agree we need to make this opt in granular.  All those places are layering problems where we're building against an LKG instead of live bits and represent problems with product layering.   We'll lose sight of those places once we have a newer SDK.

Comment (ViktorHofer) [reply src/libraries/Directory.Build.targets:295]: > I think we need it in all of Libraries in order to target 6.0. In libraries it should never cause the error (see my thread above where we set the property to disable that error).

Why do we need this in all of libraries if we already use the live built targeting pack there?

> We'll lose sight of those places once we have a newer SDK.

I agree. That said, there aren't plans yet to update to a 6.0 SDK anytime soon.

File: eng/targetingpacks.targets
Comment (ViktorHofer) [reply eng/targetingpacks.targets:7]: I extracted setting that property out into https://github.com/dotnet/runtime/pull/45858 as devs are currently blocked by it.

**Comment by ViktorHofer** [comment]
IMHO no projects in this repository should depend on a pre-built 6.0 targeting pack. If they do, then we should fix that and reference the live one by declaring a dependency to it. I understand that we need to unblock this PR but we should (immediately) follow-up on making it possible to reference the live built targeting pack with the least amount of steps. We should declare that dependency in a way that doesn't require the developer to manually build the whole libraries subset. I will file an issue.

File: src/installer/Directory.Build.targets
Comment (ericstj) [reply src/installer/Directory.Build.targets:166]: Not sure.  That's a totally different framework version: what if they need to have coherent assembly references (I don't know if this is the case or not)?  IMHO these should be building against and testing the live product instead of an LKG but that might be bigger fish to fry.  @Anipik can you comment on the extent of the dependencies on `net6.0` in installer: can we dial those back to `net5.0` without changing what we ship?

Comment (Anipik) [reply src/installer/Directory.Build.targets:166]: i am just trying that out

**Comment by ViktorHofer** [comment]
@jkoritzinsky regarding https://github.com/dotnet/runtime/pull/45781/commits/ef7adfe22555e033ff43d4bb6585c4a94215888c, do the sfxprojs need to target NetCoreAppCurrent or is any other synthetic tfm fine as well?

**Comment by jkoritzinsky** [comment]
They need to target NETCoreAppCurrent. The value goes in various data files IIRC.

File: src/installer/Directory.Build.targets
Comment (ViktorHofer) [reply src/installer/Directory.Build.targets]: to close on this, most components in installer don't need implicit framework references and hence using the latest TargetFramework works fine in these cases. `Microsoft.NET.HostModel` will be moved into the libraries section anyway. The remaining test projects presumably don't need to target a `$(NetCoreAppCurrent)` targeting pack and are fine with latest - 1.

File: src/libraries/Directory.Build.targets
Comment (safern) [reply src/libraries/Directory.Build.targets]: > Why do we need this in all of libraries if we already use the live built targeting pack there?

Because of the way you update the resolved targeting pack to point to a local directory the SDK needs to be able to resolve a `KnownFrameworkReference` for the tfm you're building.

File: src/installer/tests/Assets/TestProjects/LightupLib/LightupLib.csproj
Comment (safern) [inline src/installer/tests/Assets/TestProjects/LightupLib/LightupLib.csproj:5]: Doesn't this mean that installer tests will no longer run against live libraries bits? I don't know if they currently do. @jkoritzinsky ?

File: src/libraries/Directory.Build.targets
Comment (ViktorHofer) [reply src/libraries/Directory.Build.targets]: Sorry, I should marked the conversation as resolved. This isn't needed for all of libraries, only for parts which use framework references (anything not inbox). I fixed this with https://github.com/dotnet/runtime/commit/adad810e664f53211aed98f0f06b6beeaede180c.

File: src/installer/tests/Assets/TestProjects/LightupLib/LightupLib.csproj
Comment (jkoritzinsky) [reply src/installer/tests/Assets/TestProjects/LightupLib/LightupLib.csproj:5]: These test projects run against the live runtime. We should use NETCoreAppCurrent and restore them against the ref pack that we build in the packs.product subset.

File: src/installer/tests/Directory.Build.targets
Comment (safern) [inline src/installer/tests/Directory.Build.targets:80]: This shouldn't change either I think.

File: src/installer/tests/Assets/TestProjects/LightupLib/LightupLib.csproj
Comment (Anipik) [reply src/installer/tests/Assets/TestProjects/LightupLib/LightupLib.csproj:5]: cc @ViktorHofer

Comment (ViktorHofer) [reply src/installer/tests/Assets/TestProjects/LightupLib/LightupLib.csproj:5]: > These test projects run against the live runtime. We should use NETCoreAppCurrent and restore them against the ref pack that we build in the packs.product subset.

Master currently uses the 5.0-RC1 targeting and runtime pack: https://github.com/dotnet/runtime/blob/543cb488437ea0b05a58e3297214a0c9fb2064ea/src/installer/tests/Assets/TestUtils/TestProjects.targets#L20. If I got this right, then https://github.com/dotnet/runtime/blob/543cb488437ea0b05a58e3297214a0c9fb2064ea/src/installer/tests/Directory.Build.targets#L16 is publishing the entire bundle to disk which is then used for the test invocation.

Based on that I don't see how changing the tfm would impact whether we run against the live built runtime. I'm not an expert on the installer subset so please correct me if I'm wrong here.

**Review by ViktorHofer — APPROVED** [review]
LGTM modulo the feedback that @jkoritzinsky shared which still needs to be figured out.

File: src/installer/tests/Assets/TestProjects/LightupLib/LightupLib.csproj
Comment (jkoritzinsky) [reply src/installer/tests/Assets/TestProjects/LightupLib/LightupLib.csproj:5]: I'd prefer if we can get the targeting pack version that the projects build against to match the live built runtime so we're correctly testing what we mean to test without using rollforward.

File: src/installer/tests/Directory.Build.targets
Comment (jkoritzinsky) [inline src/installer/tests/Directory.Build.targets:137]: We only need this block with the test assets. We don't need it for the actual unit tests.
```suggestion
```

**Review by jkoritzinsky — APPROVED** [review]
LGTM with one suggested change.

File: src/installer/tests/Assets/TestProjects/LightupLib/LightupLib.csproj
Comment (ViktorHofer) [reply src/installer/tests/Assets/TestProjects/LightupLib/LightupLib.csproj]: Absolutely, we should follow-up on this. My comment was about master already using a different TargetingPackVersion that what is used during runtime which suggests that this PR shouldn't address this.

**Pull request merged**
